java/src/Ice/src/main/java/com/zeroc/Ice/ConnectionI.java

```diff
@@ -2250,56 +2286,60 @@ private int parseMessage(MessageInfo info) {
     return _state == StateHolding ? SocketOperation.None : SocketOperation.Read;
   }
 
-  private void invokeAll(
+  private void dispatchAll(
       InputStream stream,
-      int invokeNum,
+      int requestCount,
       int requestId,
       byte compress,
       com.zeroc.IceInternal.ServantManager servantManager,
       ObjectAdapter adapter) {
-    //
-    // Note: In contrast to other private or protected methods, this
-    // operation must be called *without* the mutex locked.
-    //
-
-    Incoming in = null;
-    try {
-      while (invokeNum > 0) {
 
-        //
-        // Prepare the invocation.
-        //
-        boolean response = !_endpoint.datagram() && requestId != 0;
-        in = getIncoming(adapter, response, compress, requestId);
-
-        //
-        // Dispatch the invocation.
-        //
-        in.invoke(servantManager, stream);
+    // Note: In contrast to other private or protected methods, this method must be called *without*
+    // the mutex locked.
 
-        --invokeNum;
+    Object dispatcher = adapter != null ? adapter.dispatchPipeline() : null;
 
-        reclaimIncoming(in);
-        in = null;
+    try {
+      while (requestCount > 0) {
+        // adapter can be null here, however we never pass a null current.adapter to the application
+        // code.
+        var request = new IncomingRequest(requestId, this, adapter, stream);
+        final boolean isTwoWay = !_endpoint.datagram() && requestId != 0;
+
+        if (dispatcher != null) {
+          CompletionStage<OutgoingResponse> response = null;
+          try {
+            response = dispatcher.dispatch(request);
+          } catch (RuntimeException | UserException ex) {
+            sendResponse(request.current.createOutgoingResponse(ex), isTwoWay, (byte) 0);
+          } catch (java.lang.Error ex) {
```

Comment:
We need to figure out what we do with Errors in a follow-up PR.

As of this PR, all Errors thrown during a dispatch are encoded in a response and handled (= ignored) unless synchronous and the createOutgoingResponse/sendResponse fails with a "fatal" error.